cmd/credentials.go

```diff
@@ -0,0 +1,195 @@
+		return cli.NewExitError("--provider flag missing", -1)
+	}
+
+	productLabel := cliCtx.String("product")
+	if productLabel == "" {
+		return cli.NewExitError("--product flag missing", -1)
+	}
+
+	p := promptui.Prompt{
+		Label: "Email",
+		Validate: func(input string) error {
+			valid := govalidator.IsEmail(input)
+			if valid {
+				return nil
+			}
+
+			return errors.New("Please enter a valid email address")
+		},
+	}
+
+	email, err := p.Run()
+	if err != nil {
+		return cli.NewExitError(err.Error(), -1)
+	}
+
+	p = promptui.Prompt{
+		Label: "Password",
+		Mask:  passwordMask,
+		Validate: func(input string) error {
+			if len(input) < 8 {
+				return errors.New("Passwords must be greater than 8 characters")
+			}
+
+			return nil
+		},
+	}
+
+	password, err := p.Run()
+	if err != nil {
+		return cli.NewExitError(err.Error(), -1)
+	}
+	cfgs := []manifold.ConfigFunc{}
+
+	cfgs = append(cfgs, manifold.ForURLPattern(apiURL))
+	cfgs = append(cfgs, manifold.WithUserAgent("grafton-"+version))
+
+	client := manifold.New(cfgs...)
+
+	token, err := client.Login(ctx, email, password)
+	if err != nil {
+		return cli.NewExitError(err.Error(), -1)
+	}
+
+	cfgs = append(cfgs, manifold.WithAPIToken(token))
+	client = manifold.New(cfgs...)
+
+	var provider *manifold.Provider
+	var product *manifold.Product
+
+	provList := client.Providers.List(ctx)
+
+	for provList.Next() {
+		p, err := provList.Current()
+		if err != nil {
+			return cli.NewExitError("Fetching provider error: "+err.Error(), -1)
+		}
+
+		if p.Body.Label == providerLabel {
+			provider = p
+			break
+		}
+	}
+
+	if provider == nil {
+		return cli.NewExitError(fmt.Sprintf("Provider %q not found", providerLabel), -1)
+	}
+
+	prodList := client.Products.List(ctx, nil)
+
+	for prodList.Next() {
+		p, err := prodList.Current()
+		if err != nil {
+			return cli.NewExitError("Fetching product error: "+err.Error(), -1)
+		}
+
+		if p.Body.Label == productLabel {
+			product = p
+			break
+		}
+	}
+
+	if product == nil {
+		return cli.NewExitError(fmt.Sprintf("product %q not found", productLabel), -1)
+	}
+
+	connector, err := NewConnector(token)
+	if err != nil {
+		return cli.NewExitError("Failed to create connector client "+err.Error(), -1)
+	}
+
+	desc := "grafton rotation"
+
+	params := o_auth.NewPostCredentialsParamsWithContext(ctx)
+	body := &models.OAuthCredentialCreateRequest{
+		Description: &desc,
+		ProductID:   product.ID,
+	}
+	params.SetBody(body)
+
+	res, err := connector.OAuth.PostCredentials(params, nil)
+	if err != nil {
+		return cli.NewExitError("Failed to rotate credentials "+err.Error(), -1)
+	}
+
+	payload := res.Payload
+
+	fmt.Println("Your old credentials will expire automatically in 24 hours")
+	fmt.Println("Make sure to copy your new credentials now. You won’t be able to see this information again!")
+	fmt.Printf("Client ID: %s\n", payload.ID)
+	fmt.Printf("Secret: %s\n", payload.Secret)
+
+	return nil
+}
+
+func NewConnector(token string) (*client.Connector, error) {
```

Comment:
FWIW we have a placeholder repo at https://github.com/manifoldco/go-connector that we could use for an oag generated client, eventually.

Reply:
Should I move the code there instead? I was planning to use `go-manifold`, but hit some roadblocks with oag.

Reply:
I'd stick with what you're doing right now, and once oag can handle allof and refs, we can put this stuff in go-connector, and update grafton to use it.